### DISCUSSION TITLE
Add black outline to figure on a white background

### DISCUSSION THREAD

**Discussion by justrag** [ImageMagick/ImageMagick#2550] [Help]
There's a figure on a white (non-transparent) background - how do I add a black outline? Basically, how do I go from here:
![RoyalChampion-orig](https://user-images.githubusercontent.com/10939691/92881712-85851680-f40f-11ea-9625-fe1e1f76545b.jpg)
to here:
![RoyalChampion-outline](https://user-images.githubusercontent.com/10939691/92881714-86b64380-f40f-11ea-8192-ee7c1505f7f9.jpg)
with maybe less whitespace between the outline and the figure?

Something similar to 'halo effect' in here: http://www.imagemagick.org/Usage/masking/#outline (http://www.imagemagick.org/Usage/masking/cyclops_halo.png), but in black instead of white and much less "haloey".

**Comment by snibgo** [comment]
Your image is JPEG, so it has lossy compression, so pixels you think are white may not be. A suitable "-fuzz" is needed. Second problem: pixels within your figure are white, so we can't simply declare white pixels to be background.

This gives a rough result for black knight on white background (Windows BAT syntax):
```
%IMG7%magick ^
  knight.jpg ^
  -fill Blue -fuzz 6% -draw "color 0,0 floodfill" ^
  -fill Black -fuzz 0 +opaque Blue ^
  -fill White -opaque Blue ^
  x.png
```
You could then use any edge process to get the edge effect you want.

But the result is noisy because the input is noisy. A denoise process might help, but you really need a better source.

**Reply by justrag** [reply]
Hmmm... This seems to fill the whole figure with black:
![blackknightoutline](https://user-images.githubusercontent.com/10939691/92942998-7ffcf000-f452-11ea-9819-923f33807a50.png)

**Comment by snibgo** [comment]
Yes, as I said, "black knight on white background".

**Comment by fmw42** [comment]
Here is one way to do that using IM 6 Unix syntax:

![knight](https://user-images.githubusercontent.com/13458218/92948451-40141980-f40e-11ea-997c-909c89cb5f9a.jpg)


```
convert knight.jpg \
\( -clone 0 -fuzz 15% -fill none -draw "matte 0,0 floodfill" \
    -alpha extract -morphology edgeout disk:5 -negate \) \
-compose multiply -composite \
knight_halo.jpg

```


![knight_halo](https://user-images.githubusercontent.com/13458218/92948462-44d8cd80-f40e-11ea-9404-03b66d402648.jpg)


Syntax may be different for Windows or ImageMagick 7. So please always provide your ImageMagick version and platform/OS when asking questions.
